### ISSUE TITLE
Change visibility of AndroidSentryClientFactory context to protected and add a no-args constructor

### ISSUE THREAD

**Issue by ninniuz** [getsentry/sentry-java#713]
Currently, [AndroidSentryClientFactory's context](https://github.com/getsentry/sentry-java/blob/94862776b8825bad12a94fdeacb654ba7b2ac0aa/sentry-android/src/main/java/io/sentry/android/AndroidSentryClientFactory.java#L38) is `private` preventing any subclass to set its own specific context, having to rely on the superclass constructor.

That is fairly acceptable for most cases, but in my case I need to init Sentry in my custom `Application` class before calling `Application.attachBaseContext` so I cannot rely on `AndroidSentryClientFactory` constructor which invokes `Context.getApplicationContext()`: in my case, that results in a faulty context which will throw exceptions on any method.

I would like to be able to extend `AndroidSentryClientFactory` and leverage the default behavior but still be able to specify my own context without having the base factory call on `getApplicationContext`.

**Comment by justacodefan** [comment]
Do you have an example of when you need to initialize a SentryClient before the application context is available to you? You will be unable to send any events.

You can always re-set the stored client via:
```
Sentry.setStoredClient(new AndroidSentryClientFactory(context).createSentryClient(null));
```

**Comment by justacodefan** [comment]
I opened #714 which hopefully addresses your use-case.

**Comment by ninniuz** [comment]
> Do you have an example of when you need to initialize a SentryClient before the application context is available to you? You will be unable to send any events.
> 

A typical use-case is having a custom language selector inside the app, which needs you to create a [configuration context](https://developer.android.com/reference/android/content/Context#createConfigurationContext(android.content.res.Configuration)).

In my case I have something like this
```
public class CustomApp extends Application {
    ...
    @Override
    protected void attachBaseContext(Context base) {  
        ...
        // do some work and get a configuration context
        Context configurationContext = buildConfigurationContext(base);
        super.attachBaseContext(configurationContext);
        ...
    }
}
```
As I want to be able to detect and report errors happening inside `buildConfigurationContext` I may want to init Sentry before that method call, that is before `super.attachBaseContext` is exercised.

The `base` context is a long-lived `Context` but it is lacking a base context, that means any call to `getApplicationContext` or `getBaseContext` would return a `null` value.

I just would like to be able to leverage `AndroidSentryClientFactory` facilities being able to feed the Context I see fit.

> You can always re-set the stored client via:
> 
> ```
> Sentry.setStoredClient(new AndroidSentryClientFactory(context).createSentryClient(null));
> ```

The problem is that if I create a new instance of `AndroidSentryClientFactory` with a context obtained **before** `ContextWrapper.attachBaseContext` returns (in my example, before `buildConfigurationContext`) and try to report an issue with that context, I would get NPEs.

**Comment by ninniuz** [comment]
Closed with #714
